### PR TITLE
Add Offence Event

### PR DESCRIPTION
**Pull request by mir-nawaz**
## Description
Add Offence Event

## Motivation and Context
Add Offence event for validator stats. 

## How has this been tested?
Tested with local edgeware, edgeware, and Kusama.
Unit tests added.

## Have proper tags been added (for bug, enhancement, breaking change)?
- [x] yes

## Does this PR affect any server routes?
- [ ] yes, and they are tested: [enter the % coverage here]
- [ ] yes, and they are not tested: [enter the % coverage here]
- [ ] yes, but I did not run tests
- [x] no

### CODE REVIEW

File: src/substrate/filters/enricher.ts [hunk 1]
```diff
@@ -66,6 +67,25 @@ export async function Enrich(
         }
       }
 
+      /**
+       * Offences Events
+       */
+      case EventKind.Offence: {
+        const [ offenceKind, opaqueTimeSlot, applied ] = event.data as unknown as [ Kind, OpaqueTimeSlot, boolean ];
```
Comment (jnaviask) [inline src/substrate/filters/enricher.ts:74]: This should be `bool`, the polkadot type, not `boolean`, the js native type. You can convert from the former to the latter with `applied.isTrue`.

File: src/substrate/filters/enricher.ts [hunk 2]
```diff
@@ -66,6 +67,25 @@ export async function Enrich(
...
+        const reportIds = await api.query.offences.concurrentReportsIndex(offenceKind, opaqueTimeSlot);
+        const offenceDetails: Option<OffenceDetails>[] = await api.query.offences.reports
+          .multi(reportIds.map((reportId) => reportId.toString()));
```
Comment (jnaviask) [inline src/substrate/filters/enricher.ts:77]: You shouldn't necessarily need to convert to string here, it should accept the codec type as-is.

Also this is a fairly heavy-weight API operation -- how often do we expect these events to come in? We wont want to do this very regularly.

File: src/substrate/filters/enricher.ts [hunk 3]
```diff
@@ -66,6 +67,25 @@ export async function Enrich(
...
+        const offenders : Array<ValidatorId> = offenceDetails.map((offence) => offence.unwrap().offender[0]);
```
Comment (jnaviask) [inline src/substrate/filters/enricher.ts:78]: This `unwrap()` call can throw if `offence` is `None` -- we need to handle this error (either by filtering out null options immediately using `isSome` or by using `unwrapOr()` and filtering afterward).

File: src/substrate/filters/enricher.ts [hunk 4]
```diff
@@ -66,6 +67,25 @@ export async function Enrich(
...
+        return {
+          data: {
+            kind,
+            offenceKind: offenceKind.toString(),
+            opaqueTimeSlot: opaqueTimeSlot.toString(),
+            applied: typeof applied === 'undefined' ? true : applied,
```
Comment (jnaviask) [inline src/substrate/filters/enricher.ts:84]: This conditional should be easier to manage with the `bool` change I suggested above.

File: src/substrate/types.ts [hunk 5]
```diff
@@ -158,6 +161,17 @@ export interface ISomeOffline extends IEvent {
...
 
+/**
+ * Offences Events
+ */
+export interface IOffence extends IEvent {
+  kind: EventKind.Offence;
+  offenceKind: string;
+  opaqueTimeSlot: string;
+  applied: boolean;
+  offenders: Array<ValidatorId>
```
Comment (jnaviask) [inline src/substrate/types.ts:172]: `ValidatorId` is a polkadot type, so we should replace this with `string`, or if we wanted to be consistent, use the `AccountId` type I defined above (which is just a string).

File: test/unit/edgeware/enricher.spec.ts [hunk 6]
```diff
@@ -774,7 +785,23 @@ describe('Edgeware Event Enricher Filter Tests', () => {
...
+    const event = constructEvent([ 'offline', '10000', true, offenceDetails ],
+      'offences', [ 'Kind', 'OpaqueTimeSlot', 'boolean', 'Option<OffenceDetails>[]' ]);
+    const result = await Enrich(api, blockNumber, kind, event);
+    assert.deepEqual(result, {
+      blockNumber,
+      data: {
+        kind,
+        offenceKind: 'offline',
+        opaqueTimeSlot: '10000',
+        applied: true,
```
Comment (jnaviask) [inline test/unit/edgeware/enricher.spec.ts:800]: You should use the `constructBool()` helper call to generate the test value for `applied`.

**Review by jnaviask — CHANGES_REQUESTED** [review]
Couple of small-ish changes needed, but overall looks good!

File: src/substrate/filters/enricher.ts (hunk 2 above)
Comment (drewstone) [reply src/substrate/filters/enricher.ts:77]: Offences come in every session/era so this is not an issue.

**Review by drewstone — CHANGES_REQUESTED** [review]
We'll also need a version bump on the API to v0.2.17. Will continue to review but that was top of my mind as well for this.

File: test/unit/edgeware/enricher.spec.ts [hunk 7]
```diff
@@ -774,7 +785,24 @@ describe('Edgeware Event Enricher Filter Tests', () => {
       }
     });
   });
+  /** offences events */
+  it.only('should enrich new offence event', async () => {
+    const kind = EventKind.Offence;
+    const event = constructEvent([ 'offline', '10000', constructBool(true), offenceDetails ],
+      'offences', [ 'Kind', 'OpaqueTimeSlot', 'boolean', 'Option<OffenceDetails>[]' ]);
```
Comment (jnaviask) [inline test/unit/edgeware/enricher.spec.ts:792]: You probably don't need the 2nd and 3rd arguments here, but if you wanted to keep them then I'd make sure you replace `boolean` with `bool` to ensure it uses the correct Codec type.

**Review by jnaviask — CHANGES_REQUESTED** [review]
Just a small change to the test setup and it's good to go.

**Review by drewstone — APPROVED** [review]

**Pull request merged**